### ISSUE TITLE
Limitations of pretag.map

### ISSUE THREAD

**Issue by SanderDelden** [pmacct/pmacct#470]
Hi,

I'm currently building our pmacct environment using sfacctd and I think I'm running into some limitations of the pretag.map
The more entries that are defined in the pretag.map, the less flow are actually being outputted. With about ~500 entries in the pretag.map rougly ~25% of the flows are missing, increase this to 30k and only 1-2% is shown.

My best guess for now is that this is related to limitations of the pretag.map. I've already tried to enable 'maps_index: true' to combat this, but unfortunately this is not compatible with my pretag.map.

The first limition I ran in to is that the 'src_net' key was not supported in combination with maps indexing, I solved this by using: 'filter=' src net x.x.x.x'. This seems to solve the issue however it looks like this only supports up to 4 'set_tag' entries in the pretag map. When using 4 or less 'set_tag' entries the map is sucessfully loaded:

```
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: created index 40000000 (9 entries).
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: created index 1000 (2 entries).
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: created index 0 (8 entries).
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] map successfully (re)loaded.
```

However when using 5 or more entries I get the following:

```
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: created index 40000000 (9 entries).
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: created index 1000 (2 entries).
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: created index 0 (10 entries).
WARN ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: out of index space 0. Indexing disabled.
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: destroyed index 40000000.
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: destroyed index 1000.
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] maps_index: destroyed index 0.
INFO ( default/core ): [/etc/pmacct/mappings/ipas_pretag.map] map successfully (re)loaded.
```

The configuration of my pretag.map looks as follows:

```
! Define all netflow sending devices and label them
!
set_label=router1		ip=xx.xx.xx.250   jeq=customer_id
set_label=router2		ip=xx.xx.xx.254   jeq=customer_id
set_label=router3		ip=xx.xx.xx.243   jeq=customer_id
set_label=router4		ip=xx.xx.xx.244   jeq=customer_id
set_label=router5		ip=xx.xx.xx.255   jeq=customer_id
set_label=router6		ip=xx.xx.xx.252   jeq=customer_id
set_label=router7		ip=xx.xx.xx.246   jeq=customer_id
set_label=router8		ip=xx.xx.xx.247   jeq=customer_id
set_label=router9		ip=xx.xx.xx.240   jeq=customer_id
!
! Set tag 1 for our own AS (as we want to ignore this)
!
set_tag=1       dst_as=0                label=customer_id
!
! Set tags for customer IP addresses
!
! Customer 1
set_tag=1		src_net=1.1.1.1		label=customer_id
set_tag=1		src_net=1.1.1.2		label=customer_id
set_tag=1		src_net=1.1.1.3		label=customer_id
! Customer 2
set_tag=2		src_net=2.2.2.1		label=customer_id
set_tag=2		src_net=2.2.2.2		label=customer_id
! Customer 3
set_tag=3		src_net=3.3.3.1		label=customer_id
set_tag=3		src_net=3.3.3.2		label=customer_id
set_tag=3		src_net=3.3.3.3		label=customer_id
set_tag=3		src_net=3.3.3.4		label=customer_id
...
...
...
```

My sfacctd.conf is:

```
plugins: print[CUSTOMER]

# Global config

bgp_daemon: true
bgp_daemon_max_peers: 20
bgp_agent_map:/etc/pmacct/mappings/bgp.map
nfacctd_as_new: bgp
sfacctd_port: 5008
sfacctd_renormalize: true
maps_index: true

# CUSTOMER config

pre_tag_filter[CUSTOMER]: -1
pre_tag_map[CUSTOMER]:/etc/pmacct/mappings/ipas_pretag.map
maps_entries[CUSTOMER]: 100000
maps_row_len[CUSTOMER]: 100000
aggregate[CUSTOMER]: label, tag, dst_as, src_host, out_iface
print_output_file[CUSTOMER]: /tmp/1m_IPAS.json
print_output[CUSTOMER]: json
print_history[CUSTOMER]: 1m
print_history_roundoff[CUSTOMER]: m
print_refresh_time[CUSTOMER]: 60
plugin_pipe_size[CUSTOMER]: 10240000
plugin_buffer_size[CUSTOMER]: 302400
print_cache_entries[CUSTOMER]: 302400
print_max_writers[CUSTOMER]: 50
print_output_file_append[CUSTOMER]: true
print_trigger_exec[CUSTOMER]: /etc/pmacct/scripts/send2influx_IPAS.py
```

Running version: `sfacctd 1.7.7-git (20210124-0 (71629c4))`
Please let me know if any additional information/logging is needed.

With kind regards,
Sander

**Comment by paololucente** [comment]
Hi Sander ( @SanderDelden ),

The performance issue you are verifying is because, as the log says, it is unable to build the index for the pre_tag_map so it essentially goes ahead with the same map not indexed (expensive lookups): you may probably be able to verify the process hitting 100% CPU cycles.

The above said, i was unable to reproduce the exact same indexing error as the one you get. Probably because what you posted is an excerpt of it. If it is possible, can you land me the full map (anonymized is fine) by unicast email?

This also said, and cleared what is the issue you are currently bumping into, probably we'll be hitting one more issue: indexing will refuse to load against the 'src_net' keyword because 'src_net' could be an IP prefix (so IP address + netmask) and hashing and ranges don't really go along very well. Can you confirm that, as it looks like in your example, you are going to be using 'src_net' to match against individual IP addresses, ie. /32's? If yes, then this issue is worth looking into; otherwise i may recommend tagging flows down in your data pipeline, ie. create an enrichment layer between pmacct (collection) and InfluxDB (database).

Paolo

**Comment by SanderDelden** [comment]
Hi Paolo,

My sincere apologies for the delayed reply. Unfortunately some other things came in to my schedule and I was not able to send you a reply until now.

Sending you the full map by unicast email is no problem, where could I best send it to?

I can confirm that I will be using 'src_net' to match against individual IP addresses (/32's).

Kind regards,
Sander

**Comment by paololucente** [comment]
Hi Sander,

That would be: https://github.com/pmacct/pmacct/blob/master/AUTHORS#L6

Paolo

**Comment by SanderDelden** [comment]
Hi Paolo,

I sent an email to you on the 12th of march, just checking in if it arrived correctly. If so, no rush :) If not please let me know. If need be I can attach the files here.

Kind regards,
Sander

**Comment by paololucente** [comment]
Hi Sander,

Apologies for the radio silence. Yes, email arrived & i will follow-up soon. Although unrelated to your case, right in these days - if you see the recent commits - i am refurbishing the map indexing part for improved efficiency and better functionality. Will be in touch soon.

Paolo

**Comment by paololucente** [comment]
Hi Sander,

This should be all done now (see above commits). Tested it loading against the file you kindly sent me via email and tested it working against a handful IP addresses in my lab environment. Two notes about the file you sent me: 1) `label` is really needed only for the first entry in the series, ie. you can skip having that `label=customer_id` at every single line; that said, having it does not hurt anything; 2) tags are numerical only so, say, `set_tag=156_1` will not parse well. Switch to `set_label` instead.

Please keep me posted if you manage to give this a try.

Paolo

**Comment by SanderDelden** [comment]
Hi Paolo,

I have been testing extensively with this for the past week and everything is working perfectly! Having the <code>src_net</code> key availabe in comination with maps indexing enabled is a winner.

1) Understood, I've adjusted this to only include label in the first entry, thank you for the tip!
2) For now I've removed the <code>_x</code> for every entry, should this be need in the future we'll switch to labels, thanks!

As always thank you so much for you work/assistance on this!

Kind regards,
Sander
